File: lab-3/Graph.cs
using System;
using System.Collections.Generic;

namespace Csharp_3
{
    public class Graph
    {
        private int Vertices;
        
        private List<Int32>[] adj;
        
        public Graph(int v)
        {
            Vertices = v;
            adj = new List<Int32>[v];
            for (int i = 0; i < v; i++)
            {
                adj[i] = new List<Int32>();
            }
        }
        
        public void AddEdge(int v, int w)
        {
            adj[v].Add(w);
        }
        
        public void BFS(int s)
        {
            bool[] visited = new bool[Vertices];
            
            Queue<int> queue = new();
            visited[s] = true;
            queue.Enqueue(s);
            
            while (queue.Count != 0)
            {
                s = queue.Dequeue();
                Console.WriteLine("next->" + s);
                
                foreach (Int32 next in adj[s])
                {
                    if (!visited[next])
                    {
                        visited[next] = true;
                        queue.Enqueue(next);
                    }
                }
            }
        }
        
        public void DFS(int s)
        {
            bool[] visited = new bool[Vertices];
            
            Stack<int> stack = new();
            visited[s] = true;
            stack.Push(s);

            while (stack.Count != 0)
            {
                s = stack.Pop();
                foreach (int i in adj[s])
                {
                    if (!visited[i])
                    {
                        visited[i] = true;
                        stack.Push(i);
                    }
                }
            }
        }
    }
}

File: lab-3/Graph.java
import java.util.LinkedList;
import java.util.Queue;

class Graph {
    private int V;

    private LinkedList<Integer>[] adj;
    private Queue<Integer> queue;

    public Graph(int v) {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i) {
            adj[i] = new LinkedList<>();
        }
        queue = new LinkedList<>();
    }

    public void addEdge(int v, int w) {
        adj[v].add(w);
    }


    private void DFSUtil(int vertex, boolean nodes[]) {

        nodes[vertex] = true;
        int a;

        for (int i = 0; i < adj[vertex].size(); ++i) {
            a = adj[vertex].get(i);
            if (!nodes[a]) {
                DFSUtil(a, nodes);
            }
        }
    }

    public void DFS(int v) {
        boolean already[] = new boolean[V];
        DFSUtil(v, already);
    }

    public void BFS(int n) {

        boolean nodes[] = new boolean[V];
        int a;

        nodes[n] = true;
        queue.add(n);

        while (queue.size() != 0) {
            n = queue.poll();
            System.out.print(n + " ");

            for (int i = 0; i < adj[n].size(); ++i) {
                a = adj[n].get(i);
                if (!nodes[a]) {
                    nodes[a] = true;
                    queue.add(a);
                }
            }
        }
    }
}

